### PR TITLE
Benchmark DB Framework

### PR DESCRIPTION
**Pull request by Kbhat1**
## Describe your changes and provide context
- Adds a Benchmarking Framework for different DB Backends
- Adds a Generator Framework for exporting raw kv data from chain nodes to a file
- NOTE: Adding in more randomization/shuffling across files with chunking - pushing that soon after this
- NOTE: Adding in the read latency / througput measurements after this

## Testing performed to validate your change
- Tested on local and ec2 node benchmarking with rocksdb

### CODE REVIEW

File: benchmark/dbbackend/rocksdb.go [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+
+	fmt.Printf("Average Latency: %v\n", avgLatency)
+	fmt.Printf("P50 Latency: %v\n", utils.CalculatePercentile(latencies, 50))
+	fmt.Printf("P75 Latency: %v\n", utils.CalculatePercentile(latencies, 75))
+	fmt.Printf("P99 Latency: %v\n", utils.CalculatePercentile(latencies, 99))
+
+}
+
+func (rocksDB RocksDBBackend) BenchmarkDBRead(inputKVDir string, outputDBPath string, concurrency int) {
+	return
```
Comment (Kbhat1) [inline benchmark/dbbackend/rocksdb.go:124]: Cleaning up the read benchmarking + iteration benchmarking and wil add in follow up PR

File: benchmark/cmd/dumpkv/main.go [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+	flag.StringVar(&modules, "modules", "", "Modules to export")
+	flag.StringVar(&outputDir, "output-dir", "", "Output Directory")
+	flag.StringVar(&dbBackend, "db-backend", "", "DB Backend")
+	flag.StringVar(&rawKVInputDir, "raw-kv-input-dir", "", "Input Directory for benchmark which contains the raw kv data")
+	flag.IntVar(&version, "version", 0, "db version")
+	flag.IntVar(&concurrency, "concurrency", 1, "Concurrency while writing to db")
+	flag.IntVar(&maxRetries, "max-retries", 0, "Max Retries while writing to db")
+	flag.Parse()
+
+	if command == "" {
```
Comment (yzang2019) [inline benchmark/cmd/dumpkv/main.go:44]: We can switch to cobra cli

File: benchmark/cmd/dumpkv/main.go [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+	var levelDBDir string
+	var modules string
+	var outputDir string
+	var dbBackend string
+	var rawKVInputDir string
+	var version int
+	var concurrency int
+	var maxRetries int
+	flag.StringVar(&command, "command", "", "generate or benchmark")
+	flag.StringVar(&levelDBDir, "leveldb-dir", "", "level db dir")
```
Comment (yzang2019) [inline benchmark/cmd/dumpkv/main.go:34]: we can provide a default dir

File: benchmark/utils/utils.go [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+func AddRandomBytes(data []byte, numBytes int) []byte {
+	randomBytes := make([]byte, numBytes)
+	if _, err := rand.Read(randomBytes); err != nil {
+		panic(fmt.Sprintf("Failed to generate random bytes: %v", err))
+	}
+	return append(data, randomBytes...)
+}
+
+// NOTE: Assumes latencies is sorted
+func CalculatePercentile(latencies []time.Duration, percentile float64) time.Duration {
```
Comment (yzang2019) [inline benchmark/utils/utils.go:157]: I wonder if we can use metrics or trace to track latency and throughput

Comment (Kbhat1) [reply benchmark/utils/utils.go:153]: Yes I'm planning on adding in metrics to track this as well. @yzang2019 will be in separate PR

**Review by yzang2019 — APPROVED** [review]

**Pull request merged**